lib/ansible/modules/windows/win_chocolatey_source.ps1

```diff
@@ -0,0 +1,411 @@
+        if ($null -ne $admin_only) {
+            $admin_only = [System.Convert]::ToBoolean($admin_only.Value)
+        }
+
+        $source_info = @{
+            name = $xml_source.id
+            source = $xml_source.value
+            disabled = [System.Convert]::ToBoolean($xml_source.disabled)
+            source_username = $source_username
+            priority = $priority
+            certificate = $certificate
+            bypass_proxy = $bypass_proxy
+            allow_self_service = $allow_self_service
+            admin_only = $admin_only
+
+            source_password = $source_password
+            certificate_password = $certificate_password
+        }
+        $sources.Add($source_info) > $null
+    }
+    return ,$sources
+}
+
+Function Set-RawChocolateySourceAttribute {
+    param(
+        $choco_app,
+        $name,
+        $attribute,
+        $value
+    )
+    # for private attributes like password/certificatePassword, the only way
+    # for us to preserve the existing value when changing another attirbute is
+    # to store it manually by editing the XML
+    $choco_config_path = "$(Split-Path -Path (Split-Path -Path $choco_app.Path))\config\chocolatey.config"
+    if (-not (Test-Path -Path $choco_config_path)) {
+        Fail-Json -obj $result -message "Expecting Chocolatey config file to exist at '$choco_config_path'"
+    }
+
+    try {
+        [xml]$choco_config = Get-Content -Path $choco_config_path
+    } catch {
+        Fail-Json -obj $result -message "Failed to parse Chocolatey config file at '$choco_config_path': $($_.Exception.Message)"
+    }
+    $source = $choco_config.chocolatey.sources.GetEnumerator() | Where-Object { $_.id -eq $name }
+    $source.$attribute = $value
+    $new_xml = Get-PrettyXml -xml $choco_config
+    if (-not $check_mode) {
+        [System.IO.File]::WriteAllText($choco_config_path, $new_xml)
+    }
+}
+
+Function New-ChocolateySource {
+    [Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSAvoidUsingUserNameAndPassWordParams", "", Justification="We need to use the plaintext pass in the cmdline, also using a SecureString here doesn't make sense considering the source is not secure")]
+    [Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSAvoidUsingPlainTextForPassword", "", Justification="See above")]
+    param(
+        $choco_app,
+        $name,
+        $source,
+        $source_username,
+        $source_password,
+        [switch]$encrypted_source_password,
+        $certificate,
+        $certificate_password,
+        [switch]$encrypted_certificate_password,
+        $priority,
+        $bypass_proxy,
+        $allow_self_service,
+        $admin_only
+    )
+    # build the base arguments
+    $arguments = [System.Collections.ArrayList]@($choco_app.Path,
+        "source", "add", "--name", $name, "--source", $source
+    )
+
+    # add optional arguments from user input
+    if ($null -ne $source_username) {
```

Comment:
So turning this around here is not needed, as $source_username cannot be a collection.
I think that's why we disabled that rule in pslint, so we didn't *have* to comply with something that is futile.

Reply:
You are right it isn't needed here but I've started to get into the habit of doing this everywhere. I don't want to force other people to comply but I have started to do it myself.

Reply:
I had hoped it wouldn't have come this far, but I don't disagree. At least it's easier than second guessing or leaving it up to a static analyzer (which are still not perfect wrt. this).